Add Analog AD559x MFD driver I2C bus access

jeppenodgaard opened the pull request:
This PR renames and refactors the existing code and adds I2C support.

Commit:
drivers: ad559x: add i2c bus support

Rename ad5592 files in dts, driver and include to ad559x and add support
for I2C bus which is required for AD5593.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

Commit:
tests: drivers: build_all: add build test for ad559x i2c driver

Add build-only test of the ad559x I2C based ADC, DAC and GPIO driver.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

jeppenodgaard commented:
I only have access to AD5593 thus have not tested AD5592.
@bbilas would you be able to do this?

bbilas commented:
> I only have access to AD5593 thus have not tested AD5592. @bbilas would you be able to do this?

Unfortunately, I no longer have a device with such a converter, so I won't be able to test it.

jeppenodgaard commented:
@anangl does a review require someone to test this on an AD5592?

jeppenodgaard commented:
There are quite a lot of changes. Reviewing one commit at a time should make the review process less difficult.

bbilas reviewed (commented):
I did a quick look at this and IMO we need to tweak it a bit more. @fabiobaltieri would you take a look at it (I know it's not your area but I know you enjoy optimizing the drivers as well)?

jeppenodgaard commented on drivers/adc/Kconfig.ad559x:
[You are right](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#multiple-commits-on-a-single-pr).
My idea was to make this PR a bit more digestible. I can squash the two commits.
Would it make sense to wait until PR review is done?

jeppenodgaard commented on include/zephyr/drivers/mfd/ad559x.h:
```diff
@@ -78,8 +79,43 @@ int mfd_ad5592_read_reg(const struct device *dev, uint8_t reg, uint8_t reg_data,
+ * @retval 0 if success
+ * @retval negative errno if failure
+ */
+int mfd_ad559x_read_adc_chan(const struct device *dev, uint8_t channel, uint16_t *result);
```
I am open for suggestions!
I think I considered IC specific code in mfd_ad559x.c but it does not seem scalable.
I think I searched the code base for ideas, but I cannot remember if there was a driver that did something similar.

aasinclair replied:
Looking at the datasheet this is quite tricky for this device.
Most devices that support both I2C and SPI have near identical command sequences, so the difference are small, but for these chips the commands sequences are very different.

Could this be better handled by implementing the ad5593 as a separate driver instead?

bbilas replied:
Take a look at https://github.com/zephyrproject-rtos/zephyr/tree/main/drivers/sensor/ina23x drivers. I know there is no MFD subsys used but we share the same codebase(not much but...) for both of them. Maybe AD5592 could be adapted similarly.

jeppenodgaard replied:
@bbilas @aasinclair  I've looked at it for quite a bit now to figure out how to do it without adding either e.g. adc to mfd or mfd to adc.
The best solution I can come up with is to add a `AD559X_HAS_POINTER_BYTE_MAP` define in mfd and set it based on the bus type.
Then either have a `adc_ad559x_pbm.c` and a `adc_ad559x_no_pbm.c` files which has implement the differences or have `#ifdef` inside `adc_ad559x.c`.
I also considered adding `adc_ad5592.c` and ad5592 dts files, but that approach looks like it would require quite a bit of duplicated code in the `adc_ad5592.c` and `adc_ad5593` and potentially future ad559x ICs.

bbilas replied:
@jeppenodgaard Please prepare some draft then we will be able to review it.

fabiobaltieri commented on drivers/mfd/mfd_ad559x_i2c.c:
```diff
@@ -0,0 +1,140 @@
+static const uint8_t adc559x_reg_adc_rd_pointer = 0x40;
+static const uint8_t adc559x_reg_gpio_rd_pointer = 0x60;
+
+int mfd_ad559x_i2c_read_raw(const struct device *dev, uint8_t *val, size_t len)
```
these can be static

fabiobaltieri commented on drivers/mfd/mfd_ad559x_i2c.c:
```diff
@@ -0,0 +1,140 @@
+
+#include "mfd_ad559x.h"
+
+static const uint8_t adc559x_reg_dac_wr_pointer = 0x10;
+static const uint8_t adc559x_reg_adc_rd_pointer = 0x40;
+static const uint8_t adc559x_reg_gpio_rd_pointer = 0x60;
```
use some defines?

fabiobaltieri commented on drivers/mfd/mfd_ad559x_spi.c:
```diff
@@ -0,0 +1,167 @@
+#define AD559X_DAC_WR_MSB_BIT        BIT(15)
+#define AD559X_DAC_CHANNEL_SHIFT_VAL 12
+
+int mfd_ad559x_spi_read_raw(const struct device *dev, uint8_t *val, size_t len)
```
static

fabiobaltieri commented:
> @fabiobaltieri would you take a look at it (I know it's not your area but I know you enjoy optimizing the drivers as well)?

wait what? do I?! 🤨

bbilas commented:
> > @fabiobaltieri would you take a look at it (I know it's not your area but I know you enjoy optimizing the drivers as well)?
> 
> wait what? do I?! 🤨

Wrong person alert.

LukaszMadejGrinn commented:
@jeppenodgaard I have tested driver in your initial version (before review changes) on HW I have got and it seems to work fine.
However please not that due to HW limitations I was able to verify:
- GPIO on IO1/6 as an output only,
- DAC on IO0 only,
- ADC on IO2/3/4/5.

Nevertheless taking scope of your changes into account I think rest of IOs will also work correctly.

CC @bbilas

jeppenodgaard commented:
> @jeppenodgaard I have tested driver in your initial version (before review changes) on HW I have got and it seems to work fine. 

Thank you so much!
I might am currently trying to figure out how to refactor so e.g. adc wont be part of mfd header, so there will probably come more changes, but they should not be functional changes.

jeppenodgaard commented:
@bbilas @aasinclair I refactored the code and implemented it slightly different than I suggested by using a config flag and a mdf function to get the flag. I also squashed commits to ensure it builds. Please have a look.

jeppenodgaard commented:
Ping @bbilas

henrikbrixandersen commented:
@jeppenodgaard You need to resolve the merge conflicts.

fabiobaltieri commented:
CI fail is unrelated, @aescolar can you take a look? `west build -p -b nrf5340bsim/nrf5340/cpuapp -T samples/drivers/mbox/sample.drivers.mbox.simu` fails but `nrf5340bsim/nrf5340/cpuapp` is explicitly allowed in `samples/drivers/mbox/sample.yaml`

aescolar commented:
> CI fail is unrelated..

@fabiobaltieri thanks for the ping, issue https://github.com/zephyrproject-rtos/zephyr/issues/70566, fix is being tested

bbilas commented:
Could you move `cd42487020dd59dbd79712ea9c8aa67af57b1898` and `7475fa0fac704efe9c2b22e8b14d71d3140e4bca` into separate PR?

jeppenodgaard commented:
> Could you move `cd42487020dd59dbd79712ea9c8aa67af57b1898` and `7475fa0fac704efe9c2b22e8b14d71d3140e4bca` into separate PR?

Yes:
https://github.com/zephyrproject-rtos/zephyr/pull/70600

bbilas commented on drivers/dac/dac_ad559x.c:
```diff
@@ -0,0 +1,110 @@
+		return -EINVAL;
+	}
+
+	if (mfd_ad559x_has_pointer_byte_map(config->mfd_dev)) {
```
Can we detect the bus somehow in the `mfd_ad559x` driver? I would like to have a generic i.e `mfd_ad559x_write_reg` function that would work for both I2C and SPI variants.

jeppenodgaard replied:
`write_reg` and `write_raw` are defined in the [mfd_ad559x_transfer_function](https://github.com/jeppenodgaard/zephyr/blob/50393e142f5e65a402ffba62fb3e11059ac5f84a/drivers/mfd/mfd_ad559x.h#L35) so it works for both SPI and I2C.

The register mapping is different (for some reason) for these two ICs and the registers does not always work in the same way. See e.g.:
https://github.com/jeppenodgaard/zephyr/blob/50393e142f5e65a402ffba62fb3e11059ac5f84a/drivers/adc/adc_ad559x.c#L113

I do not think it is possible to make a struct/union mapping.

jeppenodgaard replied:
@bbilas did I misunderstand your question / suggestion?

bbilas replied:
Okay for me.

bbilas commented:
@jeppenodgaard please rebase.

jeppenodgaard commented:
> @jeppenodgaard please rebase.

Done :)
